chain-tx-enclave-next/mls/src/ciphersuite.rs

```diff
@@ -296,15 +292,88 @@ impl CipherSuite {
     }
 
     /// TODO: use generic array?
-    pub fn hash(&self, data: &[u8]) -> Vec<u8> {
+    pub fn hash(self, data: &[u8]) -> Vec<u8> {
         match self {
             CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => Sha256::digest(data).to_vec(),
         }
     }
 
-    pub fn hash_len(&self) -> usize {
+    pub fn hash_len(self) -> usize {
         match self {
             CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => 32,
         }
     }
+
+    /// spec: draft-ietf-mls-protocol.md#key-schedule
+    pub fn expand_label(
+        self,
+        secret: &[u8],
+        group_context_hash: Vec<u8>,
+        label: &str,
+        context: &[u8],
+        length: u16,
+    ) -> Result<Vec<u8>, InvalidLength> {
+        match self {
+            CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => Hkdf::<Sha256>::new(
+                None, secret,
+            )
+            .expand_label(group_context_hash, label, context, length),
+        }
+    }
+
+    pub fn secret_size(self) -> u16 {
+        match self {
+            CipherSuite::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 => 32,
+        }
+    }
+
+    pub fn derive_private_key(
```

Comment:
from Ciphersuites def?

Reply:
probably fixme or change to use https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#derivekeypair-derive-key-pair

Reply:
the problem here is that not all secrets from that "key pair expand label" may be valid, so the unmarshal may panic

Reply:
added a FIXME for now, the go version seems don't handle this currently.

Reply:
that definitions isn't longer in the spec?

Reply:
that expand_label with "key pair"?

Reply:
"This function is defined to be HPKE's `DeriveKeyPair` function.", I think it refers to the [hpke spec](https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#derivekeypair-derive-key-pair) now.

Reply:
The DeriveKeyPair function should ideally implemented in hpke library?

Reply:
the hpke library just implemented derive_keypair, we changed to use that now.